File: Samples/Verbum/how-to-embed-a-barcode-into-a-document-e20025-11.1.6-/VB/Extensions/Chart/ChartImage.vb
Imports Microsoft.VisualBasic
	Imports System
	Imports System.Collections.Generic
	Imports System.Linq
	Imports DevExpress.XtraRichEdit.Fields
	Imports DevExpress.XtraRichEdit.Model
	Imports System.IO
	Imports DevExpress.XtraCharts
	Imports System.Drawing.Imaging
	Imports DevExpress.XtraRichEdit.Utils
	Imports System.Text
Namespace BizPad

	Public Class ChartImage
		Private _ViewType As ViewType = ViewType.Pie

		Private Sub ParseTypeSwitches(ByVal switches As InstructionCollection)
			Dim type As String = switches.GetString("c")
			If type Is Nothing Then
				type = String.Empty
			End If
			Select Case type.ToLower()
				Case "bar"
					_ViewType = ViewType.Bar
				Case "line"
					_ViewType = ViewType.Line
				Case Else
					_ViewType = ViewType.Pie
			End Select
		End Sub

		Private _Width? As Integer = Nothing
		Private _Height? As Integer = Nothing

		Private Sub ParseSizeSwitches(ByVal switches As InstructionCollection)
			_Width = switches.GetNullableInt("w")
			_Height = switches.GetNullableInt("h")
		End Sub
		Private _Data As New List(Of KeyValuePair(Of String, Double))()

		Private Sub ParseDataSwitches(ByVal switches As InstructionCollection)
			Dim data As String = switches.GetString("d")
			If data Is Nothing Then
				data = String.Empty
			End If

			For Each t As String In data.Split(","c)
				Dim value As Double

				Dim pair() As String = t.Split("|"c)

				If pair.Length = 0 Then
					Continue For
				End If

				If pair.Length = 1 Then
					If (Not Double.TryParse(pair(0), value)) Then
						_Data.Add(New KeyValuePair(Of String, Double)(pair(0), 0))
					Else
						_Data.Add(New KeyValuePair(Of String, Double)(String.Empty, value))
					End If
				Else
					If pair.Length = 2 Then
						If (Not Double.TryParse(pair(1), value)) Then
							_Data.Add(New KeyValuePair(Of String, Double)(pair(0), 0))
						Else
							_Data.Add(New KeyValuePair(Of String, Double)(pair(0), value))
						End If
					End If
				End If
			Next t
		End Sub

		Private _ShowLegent As Boolean = False

		Private Sub ParseLegentSwitches(ByVal switches As InstructionCollection)
			_ShowLegent = switches.GetBool("l")
		End Sub

		Public Sub Initialize(ByVal switches As InstructionCollection)
			ParseDataSwitches(switches)
			ParseSizeSwitches(switches)
			ParseTypeSwitches(switches)
			ParseLegentSwitches(switches)
		End Sub

		Public Function CreateChart() As Stream

			Using chart As New ChartControl()
				If _Width.HasValue Then
					chart.Width = _Width.Value
				End If
				If _Height.HasValue Then
					chart.Height = _Height.Value
				End If

				Dim undefined As Integer = 1
				Dim stream As New MemoryStream()
				Try
					Dim series As New Series("Chart", _ViewType)
					Try
						If TypeOf series.Label Is DevExpress.XtraCharts.PieSeriesLabel Then
							CType(series.Label, DevExpress.XtraCharts.PieSeriesLabel).Position = PieSeriesLabelPosition.Inside
						End If

                        If _ViewType = ViewType.Pie Then
                            series.PointOptions.ValueNumericOptions.Format = DevExpress.XtraCharts.NumericFormat.Percent
                        End If

						If _Data Is Nothing OrElse _Data.Count = 0 Then
							series.Points.Add(New SeriesPoint("Undefined", New Double() { 1 }))
							series.PointOptions.PointView = PointView.SeriesName

						Else
							series.PointOptions.PointView = PointView.ArgumentAndValues

							For i As Integer = 0 To _Data.Count - 1

								Dim argument As String = _Data(i).Key.Trim()

								If String.IsNullOrEmpty(argument) Then
									argument = "Undefined " & undefined
									undefined += 1
								End If

								series.Points.Add(New SeriesPoint(argument, New Double() { _Data(i).Value }))
							Next i
						End If

						chart.Legend.Visible = _ShowLegent
						chart.BorderOptions.Visible = False

						chart.Series.AddRange(New Series() { series })
						series = Nothing

						chart.ExportToImage(stream, ImageFormat.Bmp)

						Return stream

					Catch
						If series IsNot Nothing Then
							series.Dispose()
						End If
						Throw
					End Try

				Catch
					If stream IsNot Nothing Then
						stream.Dispose()
					End If
					Throw
				End Try
			End Using
		End Function

		Public Function CreateImage() As RichEditImage
			Return RichEditImage.CreateImage(CreateChart())
		End Function

	End Class
End Namespace

File: Samples/Verbum/how-to-embed-a-barcode-into-a-document-e20025-11.1.6-/CS/Extensions/Chart/ChartImage.cs
namespace BizPad {
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using DevExpress.XtraRichEdit.Fields;
    using DevExpress.XtraRichEdit.Model;
    using System.IO;
    using DevExpress.XtraCharts;
    using System.Drawing.Imaging;
    using DevExpress.XtraRichEdit.Utils;
    using System.Text;

    public class ChartImage {
        ViewType _ViewType = ViewType.Pie;

        void ParseTypeSwitches(InstructionCollection switches) {
            string type = switches.GetString("c");
            if (type == null)
                type = String.Empty;
            switch (type.ToLower()) {
                case "bar":
                    _ViewType = ViewType.Bar;
                    break;
                case "line":
                    _ViewType = ViewType.Line;
                    break;
                case "pie":
                default:
                    _ViewType = ViewType.Pie;
                    break;
            }
        }

        Nullable<int> _Width = null;
        Nullable<int> _Height = null;

        void ParseSizeSwitches(InstructionCollection switches) {
            _Width = switches.GetNullableInt("w");
            _Height = switches.GetNullableInt("h");
        }
        List<KeyValuePair<string, double>> _Data = new List<KeyValuePair<string, double>>();

        void ParseDataSwitches(InstructionCollection switches) {
            string data = switches.GetString("d");
            if (data == null)
                data = String.Empty;

            foreach (string t in data.Split(',')) {
                double value;

                string[] pair = t.Split('|');

                if (pair.Length == 0) {
                    continue;
                }

                if (pair.Length == 1) {
                    if (!double.TryParse(pair[0], out value)) {
                        _Data.Add(new KeyValuePair<string, double>(pair[0], 0));
                    } else {
                        _Data.Add(new KeyValuePair<string, double>(String.Empty, value));
                    }
                } else
                    if (pair.Length == 2) {
                        if (!double.TryParse(pair[1], out value)) {
                            _Data.Add(new KeyValuePair<string, double>(pair[0], 0));
                        } else {
                            _Data.Add(new KeyValuePair<string, double>(pair[0], value));
                        }
                    }
            }
        }

        bool _ShowLegent = false;

        void ParseLegentSwitches(InstructionCollection switches) {
            _ShowLegent = switches.GetBool("l");
        }

        public void Initialize(InstructionCollection switches) {
            ParseDataSwitches(switches);
            ParseSizeSwitches(switches);
            ParseTypeSwitches(switches);
            ParseLegentSwitches(switches);
        }

        public Stream CreateChart() {

            using (ChartControl chart = new ChartControl()) {
                if (_Width.HasValue) {
                    chart.Width = _Width.Value;
                }
                if (_Height.HasValue) {
                    chart.Height = _Height.Value;
                }

                int undefined = 1;
                MemoryStream stream = new MemoryStream();
                try {
                    Series series = new Series("Chart", _ViewType);
                    try {
                        if (series.Label is DevExpress.XtraCharts.PieSeriesLabel) {
                            ((DevExpress.XtraCharts.PieSeriesLabel)series.Label).Position = PieSeriesLabelPosition.Inside;
                        }

                        if (_ViewType == ViewType.Pie) {
                            series.PointOptions.ValueNumericOptions.Format = DevExpress.XtraCharts.NumericFormat.Percent;
                        }

                        if (_Data == null || _Data.Count == 0) {
                            series.Points.Add(new SeriesPoint("Undefined", new double[] { 1 }));
                            series.PointOptions.PointView = PointView.SeriesName;

                        } else {
                            series.PointOptions.PointView = PointView.ArgumentAndValues;

                            for (int i = 0; i < _Data.Count; i++) {

                                string argument = _Data[i].Key.Trim();

                                if (String.IsNullOrEmpty(argument)) {
                                    argument = "Undefined " + undefined;
                                    undefined++;
                                }

                                series.Points.Add(new SeriesPoint(argument, new double[] {
                                    _Data[i].Value }));
                            }
                        }

                        chart.Legend.Visible = _ShowLegent;
                        chart.BorderOptions.Visible = false;

                        chart.Series.AddRange(new Series[] { series });
                        series = null;

                        chart.ExportToImage(stream, ImageFormat.Bmp);

                        return stream;

                    } catch {
                        if (series != null) {
                            series.Dispose();
                        }
                        throw;
                    }

                } catch {
                    if (stream != null) {
                        stream.Dispose();
                    }
                    throw;
                }
            }
        }

        public RichEditImage CreateImage() {
            return RichEditImage.CreateImage(CreateChart());
        }

    }
}
